### DISCUSSION TITLE
Why does the Composite Encoder/Layout configuration not work ?

### DISCUSSION THREAD

**Discussion by zhangzhen0214** [logfellow/logstash-logback-encoder#920] [Q&A] [answered]
[Composite Encoder/Layout](https://github.com/logfellow/logstash-logback-encoder/blob/main/README.md#composite-encoderlayout) According to the configuration in the documentation link, but the stack trace does not take effect. 
My env info:

> - springboot.version : 2.7.6
> - logback-core/classic/access : 1.2.11
> - logstash-logback-encoder : 7.2

My logback.xml is as follows:

> ```<?xml version="1.0" encoding="UTF-8" ?>
> <configuration scan="true" scanPeriod="2 seconds" debug="false">
> <appender name="CONSOLE" class="ch.qos.logback.core.ConsoleAppender">
>         <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
>             <providers>
>                 <timestamp/>
>                 <logLevel/>
>                 <threadName/>
>                 <loggerName/>
>                 <mdc/>
>                 <message/>
>                 <stackTrace>
>                     <throwableConverter class="net.logstash.logback.stacktrace.ShortenedThrowableConverter">
>                         <maxDepthPerThrowable>20</maxDepthPerThrowable>
>                         <shortenedClassNameLength>30</shortenedClassNameLength>
>                         <maxLength>1000</maxLength>
>                         <rootCauseFirst>true</rootCauseFirst>
>                         </throwableConverter>
>                 </stackTrace>
>         </encoder>
> </appender>
> <root>
>     <level value="INFO" />
>     <appender-ref ref="CONSOLE" />
> </root>
> </configuration>
Local debug, class ShortenedThrowableConverter didn't get to the corresponding configuration value (for example: maxDepthPerThrowable shortenedClassNameLength and maxLength) 
![图片](https://user-images.githubusercontent.com/23568880/215435065-d950088c-f9e2-4126-8fee-a86abbcc5ca4.png)

**Comment by philsttr** [comment, marked as answer]
The `parseOptions` method is only relevant when parsing options from a string format when used by the pattern layout.  For example:

```xml
    <conversionRule conversionWord="stack" converterClass="net.logstash.logback.stacktrace.ShortenedThrowableConverter" />

    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder>
            <pattern>[%thread] - %msg%n%stack{5,1024,10,rootFirst,regex1,regex2,evaluatorName}</pattern>
        </encoder>
    </appender>
 ```

When used within a `LoggingEventCompositeJsonEncoder` (just like your example), the options are set directly via the setter methods from logback's xml configuration logic.  In other words, `parseOptions` is not taken into consideration in this case.

I tried your configuration file, and all the options were configured properly, but I had to add a missing `</providers>` end tag.

When stopping at a breakpoint as shown in your image, the options will have already been set by logback.  You can inspect the `ShortenedThrowableConverter` object field values to confirm.

**Reply by zhangzhen0214** [reply]
Thanks for your answer，you are right, it was my omission the end tag, actually there is.
